### ISSUE TITLE
Part 3: ADD support of training regression and binary classification to the Trainer class

### ISSUE THREAD

**Issue by nzarif** [NVIDIA-Merlin/Transformers4Rec#554]
Tasks included in this part:

- Analyze and verify the logical correctness of calling `calculate_metrics` from `Model.fit()` at this [line](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/4b1a6f1dfa99d4c5c6fe2b525d5ddae3d52103ff/transformers4rec/torch/model/base.py#L676).

- Add a simple script for evaluation of metrics and results when using `Trainer` for `BinaryClassification` and `Regression` such that they can be used for CI purposes like [this one](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/4b1a6f1dfa99d4c5c6fe2b525d5ddae3d52103ff/ci/test_integration.sh) that is being used for `NextItemPrediction`.
- Version control fixes:
_ Rebase with refactor_part2 and main to get synthetic music data testing for BC and regression.
_ Pull the changes of fix-inference branch to get latest `predict_step` and the updated ci tes
- Update the testing/training flags inside the `predict_step` and `eval_loop` and test they are working correctly.
- Move the top-k logic (specific to the next item prediction task) to a function of the Trainer class and call it inside the eval_loop only when the item-prediction task is used.
- Customize train_step to get the targets for the predictions tasks:
_ iterate over the dataloader in [training_loop](https://github.com/huggingface/transformers/blob/afce73bd9d891b55dcb8d4d875d17718ffa01ff0/src/transformers/trainer.py#L1769) using input that contains labels/targets added to inputs.
_ Call [training_step](https://github.com/huggingface/transformers/blob/afce73bd9d891b55dcb8d4d875d17718ffa01ff0/src/transformers/trainer.py#L2528) inside training loop and make sure inputs encompassed with labels are passed to it.
_ Call [compute_loss()](https://github.com/huggingface/transformers/blob/afce73bd9d891b55dcb8d4d875d17718ffa01ff0/src/transformers/trainer.py#L2576) from [training_step()](https://github.com/huggingface/transformers/blob/afce73bd9d891b55dcb8d4d875d17718ffa01ff0/src/transformers/trainer.py#L2528) and make sure labels arrive there with inputs.
_ Make sure the [model()](https://github.com/huggingface/transformers/blob/afce73bd9d891b55dcb8d4d875d17718ffa01ff0/src/transformers/trainer.py#L2586) is called with correct arguments passed to it.
- Use HF [nested_detach()](https://github.com/huggingface/transformers/blob/335f9bcd3457d2c4c9c6d9aeb2dae06ab20a7b56/src/transformers/trainer_pt_utils.py#L162) inside [T4Rec.Trainer.prediction_step()](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/ec92eb6500dad7a8e6222523ff54f89c80e196e6/transformers4rec/torch/trainer.py#L340) to handle cases when we have dict instead of Tensor. This will make sure corresponding inputs and targets are placed on the same device when using multiple GPUs.
- Add a unit test for testing the trained with binary and/or regression tasks.
- Explain the support of BC and regression tasks in the documentation of the Trainer class.
